File: 200/238 - Product of Array Except Self.java
public class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ret = new int[nums.length];
        int i, temp;
        
        ret[0] = 1;
        for (i = 1; i < nums.length; ++i)
            ret[i] = ret[i - 1] * nums[i - 1];
        
        temp = 1;
        for (i = nums.length - 2; i >= 0; --i) {
            temp *= nums[i + 1];
            ret[i] *= temp;
        }
        
        return ret;
    }
}

File: csharp/238.product-of-array-except-self.cs
/*
 * @lc app=leetcode id=238 lang=csharp
 *
 * [238] Product of Array Except Self
 */
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int n = nums.Length, right = 1;
        int[] ret = new int[n];
        ret[0] = 1;
        for (int i = 1; i < n; ++i) {
            ret[i] = ret[i - 1] * nums[i - 1];
        }
        for (int i = n - 1; i >= 0; --i) {
            ret[i] *= right;
            right *= nums[i];
        }
        return ret;
    }
}

